cmd/server.go

```diff
@@ -0,0 +1,352 @@
+	// create trace directory
+	outpath := filepath.Join(serverConf.outDir, span.TraceID)
+	os.Mkdir(outpath, 0755) // ignore errors for now
+
+	// create span directory
+	outpath = filepath.Join(outpath, span.SpanID)
+	os.Mkdir(outpath, 0755) // ignore errors for now
+
+	// write span to file
+	sjs, _ := json.Marshal(span)
+	spanfile := filepath.Join(outpath, "span.json")
+	err := os.WriteFile(spanfile, sjs, 0644)
+	if err != nil {
+		log.Fatalf("could not write span json to %q: %s", spanfile, err)
+	}
+
+	// only write events out if there is at least one
+	if len(events) > 0 {
+		ejs, _ := json.Marshal(events)
+		eventsfile := filepath.Join(outpath, "events.json")
+		err = os.WriteFile(eventsfile, ejs, 0644)
+		if err != nil {
+			log.Fatalf("could not write span events json to %q: %s", eventsfile, err)
+		}
+	}
+
+	if serverConf.verbose {
+		log.Printf("[%d] wrote trace id %s span id %s to %s\n", cs.spansSeen, span.TraceID, span.SpanID, spanfile)
+	}
+}
+
+// drawPterm takes the given span and events, appends them to the in-memory
+// event list, sorts that, then prints it as a pterm table.
+func (cs *cliServer) drawPterm(span CliEvent, events []CliEvent) {
+	cs.events = append(cs.events, span)
+	cs.events = append(cs.events, events...)
+	sort.Sort(cs.events)
+	cs.trimEvents()
+
+	td := pterm.TableData{
+		{"Trace ID", "Span ID", "Parent", "Name", "Kind", "Start", "End", "Elapsed"},
+	}
+
+	top := cs.events[0] // for calculating time offsets
+	for _, e := range cs.events {
+		// if the trace id changes, reset the top event used to calculate offsets
+		if e.TraceID != top.TraceID {
+			// make sure we have the youngest possible (expensive but whatever)
+			// TODO: figure out how events are even getting inserted before a span
+			top = e
+			for _, te := range cs.events {
+				if te.TraceID == top.TraceID && te.nanos < top.nanos {
+					log.Println("SWITCHED YER TOP")
+					top = te
+					break
+				}
+			}
+		}
+
+		var startOffset, endOffset string
+		if e.Kind == "event" {
+			e.TraceID = ""
+			e.SpanID = ""
+			startOffset = strconv.FormatInt(e.Start.Sub(top.Start).Milliseconds(), 10)
+		} else {
+			if e.TraceID == top.TraceID && e.SpanID != top.SpanID {
+				e.TraceID = "" // hide it after printing the first trace id
+			}
+			so := e.Start.Sub(top.Start).Milliseconds()
+			startOffset = strconv.FormatInt(so, 10)
+			eo := e.End.Sub(top.Start).Milliseconds()
+			endOffset = strconv.FormatInt(eo, 10)
+		}
+
+		td = append(td, []string{
+			e.TraceID,
+			e.SpanID,
+			e.Parent,
+			e.Name,
+			e.Kind,
+			startOffset,
+			endOffset,
+			strconv.FormatInt(e.ElapsedMs, 10),
+		})
+	}
+
+	cs.area.Update(pterm.DefaultTable.WithHasHeader().WithData(td).Srender())
+}
+
+// trimEvents looks to see if there's room on the screen for the number of incoming
+// events and removes the oldest traces until there's room
+// TODO: how to hand really huge traces that would scroll off the screen entirely?
```

Comment:
> // TODO: how to hand really huge traces that would scroll off the screen entirely?

Honestly I would just allow the terminal to scroll, rather than scrolling the viewport for the user. If I were using this to debug something, I would expect that I could scroll back in my tmux buffer to see the previous traces, but right now that just ends up looking like this when I try:

<img width="1439" alt="Screen Shot 2021-08-18 at 8 24 45 AM" src="https://user-images.githubusercontent.com/1781907/129906047-48caa698-f100-464c-838e-b061ad04a02f.png">

As a bonus, just letting it scroll in the terminal also lets you simplify the code. 😄

Reply:
> As a bonus, just letting it scroll in the terminal also lets you simplify the code. 😄

Perhaps that's a reason to have two modes, e.g. `--stdout` that "just prints" and leaves it up to the user to paginate, columnize (?), etc.

🤔 Though, I can see myself getting there by writing to `/dev/stdout` as a file. Perhaps it's a question of "format" then? Sometimes I want text going somewhere (terminal, scrollback). Sometimes I want JSON going somewhere. Sometimes I want something like top/watch that always fits on a single screen.

Reply:
Good thoughts. I'm gonna rework --json-out and --pterm into subcommands (and separate files) so you could `--json-out -` and have it print json lines to stdout and do whatever.

I'll do that split today so we can evolve how they work separately. I agree it should just scroll or at least have a mode for that. I also kinda want to try doing more of a rich TUI but that can be a different mode, later.